docs/src/design/many_differentials.md

```diff
@@ -0,0 +1,176 @@
+
+### Structural differential types
+AD cannot automatically determine natural differential types for a primal. For some types we may be able to declare manually their natural differential type. Other types will not have natural differential types at all - e.g. `NamedTuple`, `Tuple`, `WebServer`, `Flux.Dense` -  so we are destined to make some up.
+So beyond _natural_ differential types, we also have _structural_ differential types.
+ChainRules uses `Composite{P, <:NamedTuple}` to represent a structural differential type corresponding to primal type `P`.
+`Zygote v0.4` uses `NamedTuple`.
+Structural differentials are derived from the structure of the input.
+Either automatically, as part of the AD, or manually as part of a custom rule.
+
+Consider the structure of `DateTime`:
+```julia
+julia> dump(now())
+DateTime
+  instant: UTInstant{Millisecond}
+    periods: Millisecond
+      value: Int64 63719890305605
+```
+
+The corresponding structural differential is;
+```julia
+Composite{DateTime}(
+    instant::Composite{UTInstant{Millisecond}}(
+        periods::Composite{Millisecond}(
+                value::Int64
+        )
+    )
+)
+
+```
+
+!!! note “One must be allowed to take derivatives of integer arguments”
+    This brings up another contrast to Swift.
+    In Swift `Int` is considered non-differentiable, which is quite reasonable; it doesn’t have a very good definition of the limit of a small step (as that would be some floating/fixed point type).
+    `Int` is intrinsically discrete.
+    It is commonly used for indexing, and if one takes a gradient step, say turning `x[2]` into `x[2.1]` then that is an error.
+    However, disallowing `Int` to be used as a differential means we can not handle cases like `DateTime` having a inner field of milliseconds counted as an integer from the unix epoch; or other cases where an integer is used as a convenience for computational efficiency.
+    In the case where a custom sensitivity rule claims that there is a non-zero derivative for an `Int` argument that is being used for indexing, that code is simply wrong.
+    We can’t handle incorrect code and trying to is a path toward madness.
+    Julia, unlike Swift, is not well suited to handling rules about what you can and can’t do with particular types.
+
+So the structural differential is another type of differential.
+Since AD can only create structural differentials, unless using custom sensitivity rules; and since all custom sensitivities are only written in terms of natural differentials, as that is what is used in papers about derivatives.
+So you need to support both.
+
+### Semi-structural differentials
+Where there is no natural differential type for the outermost type, but there is for some of its fields, which can call this a semi-structural differential.
```

Comment:
This section uses _semi-structural differential_ for two related things.
Both of them use the types that exist for structural differentials, but do not entirely follow the structure of the type.
Both show up in human written rules only.

I wonder if we should use two words.

The first kind is where you use a structural differential type (`Composite`) for the outer most types but natural for (some) of the inside types.
This could be called mixed-structural or something.

The second is where you use something like structural but based on the propertynames, rather than the fieldnames.
This could be called _property-structural_ ?
 
This could be called a